phylogeny/prototypeSelection.py

```diff
@@ -201,3 +202,173 @@ def prototype_selection_constructive_maxdist(dm: DistanceMatrix,
+
+       Unfortunately, we need to vary epsilon such that the desired number of
+       prototypes is found. It is very likely that we cannot find the optimal
+       value for epsilon and we try at most <steps> times.
+
+    Parameters
+    ----------
+    dm: skbio.stats.distance.DistanceMatrix
+        Pairwise distances for all elements in the full set S.
+    num_prototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+    steps: int
+        Maximal number of steps used to find a suitable epsilon.
+
+    Returns
+    -------
+    list of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
+        elements in the distance matrix.
+
+    Raises
+    ------
+    RuntimeError
+        There is a very naive optimization strategy in place to find a suitable
+        epsilon to find the desired number of prototypes. It will abort after
+        <steps> unsuccessful tries. Default is 100.
+    ValueError
+        The number of prototypes to be found should be at least 2 and at most
+        one element smaller than elements in the distance matrix. Otherwise, a
+        ValueError is raised.
+
+    Notes
+    -----
+    Timing: %timeit -n 100 prototype_selection_constructive_protoclass(dm, 100)
+            10 loops, best of 3: 32.8 s per loop
+            where the dm holds 27,398 elements
+    function signature with type annotation for future use with python >= 3.5:
+    def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
+    num_prototypes: int, steps=100: int) -> List[str]:
+    '''
+    if num_prototypes < 2:
+        raise ValueError(("'num_prototypes' must be >= 2, since a single "
+                          "prototype is useless."))
+    if num_prototypes >= len(dm.ids):
+        raise ValueError(("'num_prototypes' must be smaller than the number of"
+                          " elements in the distance matrix, otherwise no "
+                          "reduction is necessary."))
+
+    # this function is basically a search for a suitable epsilon and wraps
+    # the protoclass function
+
+    # initiate epsilon with a more or less arbitrary value
+    epsilon = dm.data.mean()
+    # define how much epsilon should be changes in an interation
+    stepSize = 0.2
+
+    # must epsilon be increased (+1) or decreased (-1). Initiallz no change
+    # take place, thus the direction is neutral.
+    direction = 0
+    # list of prototypes
+    prototypes = []
+    for i in range(steps):
+        oldDirection = direction
+        oldEpsilon = epsilon
+
+        # increase the stepsize in each iteration to converge faster
+        stepSize *= 1.1
+        # call the protoclass with a defined epsilon
+        prototypes = _protoclass(dm, epsilon)
+        # check if direction of epsilon changes has changed
```

Comment:
The whole method is for finding an epsilon value for which len(prototypes) == num_prototypes. Thus it does matter and indicates the successful end of the search. See a few lines below.